### PR TITLE
fix/test/docs: update variable sizes, add namespace in graph/depth_first_search_with_stack.cpp

### PR DESCRIPTION
**Pull request by sauravUppoor**
#### Description of Change

Fixes:

- Added namespace `graph`
- Generalised variable data types from `int` to `size_t`
- Added `addEdge()` function for uniformity with other graph algorithms.

Tests:

- Added 3 test cases.

Docs:

- Added documentation throughout the code.

#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: 

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C-Plus-Plus/pull/1335"></a>

### CODE REVIEW

File: graph/depth_first_search_with_stack.cpp [hunk 1]
```diff
@@ -1,46 +1,198 @@
...
     }
+    return traversed_path;
 }
+} /// namespace graph
+
+/**
+ * Self-test implementation
+ * @returns none
+ */
+void tests() {
```
Comment (Panquesito7) [inline graph/depth_first_search_with_stack.cpp:114]:
```suggestion
static void tests() {
```

File: graph/depth_first_search_with_stack.cpp [hunk 2]
```diff
@@ -1,46 +1,198 @@
...
                     checked[it] = GREY;
                 }
             }
-            checked[act] = BLACK;  // nodo controllato
+            checked[act] = BLACK;  /// Node has been explored
         }
     }
+    return traversed_path;
 }
+} /// namespace graph
```
Comment (Panquesito7) [inline graph/depth_first_search_with_stack.cpp:108]:
```suggestion
}  // namespace depth_first_search
}  // namespace graph
```

File: graph/depth_first_search_with_stack.cpp [hunk 3]
```diff
@@ -1,46 +1,198 @@
...
+constexpr int64_t INF = std::numeric_limits<int16_t>::max();
 
-constexpr int WHITE = 0;
-constexpr int GREY = 1;
-constexpr int BLACK = 2;
-constexpr int INF = 99999;
 
-void dfs(const std::vector<std::list<int> > &graph, int start) {
-    std::vector<int> checked(graph.size(), WHITE);
+namespace graph {
```
Comment (Panquesito7) [inline graph/depth_first_search_with_stack.cpp:45]:
```suggestion
/**
 * @namespace graph
 * @brief Graph algorithms
 */
namespace graph {
/**
 * @namespace depth_first_search
 * @brief Functions for [Depth First Search](https://en.wikipedia.org/wiki/Depth-first_search) algorithm
 */
namespace depth_first_search {
```

File: graph/depth_first_search_with_stack.cpp [hunk 4]
```diff
@@ -1,46 +1,198 @@
...
 int main() {
-    int n = 0;
-    std::cin >> n;
-    std::vector<std::list<int> > graph(INF);
-    for (int i = 0; i < n; ++i) {
-        int u = 0, w = 0;
-        std::cin >> u >> w;
-        graph[u].push_back(w);
-    }
+    tests();
```
Comment (Panquesito7) [inline graph/depth_first_search_with_stack.cpp:165]:
```suggestion
    tests();  // execute the tests
```

File: graph/depth_first_search_with_stack.cpp [hunk 5]
```diff
@@ -1,46 +1,198 @@
...
+	size_t start_pos;
+
+	/// Test 1
+	std::cout << "Case 1: " << std::endl;
+	start_pos = 1;
+	std::vector<std::vector<size_t> > g1(3, std::vector<size_t>());
+
+	graph::addEdge(&g1, 1, 2);
+	graph::addEdge(&g1, 2, 3);
+	graph::addEdge(&g1, 3, 1);
```
Comment (Panquesito7) [inline graph/depth_first_search_with_stack.cpp:124]:
```suggestion
	graph::depth_first_search::addEdge(&g1, 1, 2);
	graph::depth_first_search::addEdge(&g1, 2, 3);
	graph::depth_first_search::addEdge(&g1, 3, 1);
```

File: graph/depth_first_search_with_stack.cpp [hunk 6]
```diff
@@ -1,46 +1,198 @@
...
+
+	std::vector<size_t> expected1 {1, 2, 3}; /// for the above sample data, this is the expected output
+	assert(graph::dfs(g1, start_pos - 1) == expected1);
```
Comment (Panquesito7) [inline graph/depth_first_search_with_stack.cpp:127]:
```suggestion
	assert(graph::depth_first_search::dfs(g1, start_pos - 1) == expected1);
```

File: graph/depth_first_search_with_stack.cpp [hunk 7]
```diff
@@ -1,46 +1,198 @@
...
+
+	/// Test 2
+	std::cout << "Case 2: " << std::endl;
+	start_pos = 1;
+	std::vector<std::vector<size_t> > g2(4, std::vector<size_t>());
+
+	graph::addEdge(&g2, 1, 2);
+	graph::addEdge(&g2, 1, 3);
+	graph::addEdge(&g2, 2, 4);
+	graph::addEdge(&g2, 4, 1);
```
Comment (Panquesito7) [inline graph/depth_first_search_with_stack.cpp:138]:
```suggestion
	graph::depth_first_search::addEdge(&g2, 1, 2);
	graph::depth_first_search::addEdge(&g2, 1, 3);
	graph::depth_first_search::addEdge(&g2, 2, 4);
	graph::depth_first_search::addEdge(&g2, 4, 1);
```

File: graph/depth_first_search_with_stack.cpp [hunk 8]
```diff
@@ -1,46 +1,198 @@
...
 
+	std::vector<size_t> expected2 {1, 3, 2, 4}; /// for the above sample data, this is the expected output
+	assert(graph::dfs(g2, start_pos - 1) == expected2);
```
Comment (Panquesito7) [inline graph/depth_first_search_with_stack.cpp:141]:
```suggestion
	assert(graph::depth_first_search::dfs(g2, start_pos - 1) == expected2);
```

File: graph/depth_first_search_with_stack.cpp [hunk 9]
```diff
@@ -1,46 +1,198 @@
...
+
+	/// Test 3
+	std::cout << "Case 3: " << std::endl;
+	start_pos = 2;
+	std::vector<std::vector<size_t> > g3(4, std::vector<size_t>());
+
+	graph::addEdge(&g3, 1, 2);
+	graph::addEdge(&g3, 1, 3);
+	graph::addEdge(&g3, 2, 4);
+	graph::addEdge(&g3, 4, 1);
```
Comment (Panquesito7) [inline graph/depth_first_search_with_stack.cpp:152]:
```suggestion
	graph::depth_first_search::addEdge(&g3, 1, 2);
	graph::depth_first_search::addEdge(&g3, 1, 3);
	graph::depth_first_search::addEdge(&g3, 2, 4);
	graph::depth_first_search::addEdge(&g3, 4, 1);
```

File: graph/depth_first_search_with_stack.cpp [hunk 10]
```diff
@@ -1,46 +1,198 @@
...
+
+	std::vector<size_t> expected3 {2, 4, 1, 3}; /// for the above sample data, this is the expected output
+	assert(graph::dfs(g3, start_pos - 1) == expected3);
```
Comment (Panquesito7) [inline graph/depth_first_search_with_stack.cpp:155]:
```suggestion
	assert(graph::depth_first_search::dfs(g3, start_pos - 1) == expected3);
```

File: graph/depth_first_search_with_stack.cpp [hunk 11]
```diff
@@ -1,46 +1,198 @@
...
+
+    /// creating a graph
+    std::vector<std::vector<size_t> > adj(vertices, std::vector<size_t>());
+
+    /// taking input for the edges
+    std::cout << "Enter the vertices which have edges between them : " << std::endl;
+	while (edges--) {
+		size_t u = 0, v = 0;
+		std::cin >> u >> v;
+		graph::addEdge(&adj, u, v);
```
Comment (Panquesito7) [inline graph/depth_first_search_with_stack.cpp:183]:
```suggestion
		graph::depth_first_search::addEdge(&adj, u, v);
```

File: graph/depth_first_search_with_stack.cpp [hunk 12]
```diff
@@ -1,46 +1,198 @@
...
+	}
+
+    /// taking input for the starting position
+    std::cout << "Enter the starting vertex [1,n]: " << std::endl;
+	std::cin >> start_pos;
+	start_pos -= 1;
+	traversal = graph::dfs(adj, start_pos);
```
Comment (Panquesito7) [inline graph/depth_first_search_with_stack.cpp:190]:
```suggestion
	traversal = graph::depth_first_search::dfs(adj, start_pos);
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
👍 Well done.

**Comment by sauravUppoor** [comment]
@Panquesito7 Thanks a lot for going through the code. I have committed the suggested changes. 😄

File: graph/depth_first_search_with_stack.cpp [hunk 13]
```diff
@@ -1,46 +1,208 @@
...
+ * 4. Push all its neighbours which are not coloured BLACK.
+ * 5. Repeat steps 4 and 5 until the stack is empty.
+ */
+
 #include <iostream>
 #include <list>
 #include <stack>
 #include <vector>
+#include <cassert>
+#include <limits>
```
Comment (Panquesito7) [inline graph/depth_first_search_with_stack.cpp:37]: Add a brief description of what the header/library add (see the examples below).
```suggestion
#include <iostream>  /// for io operations
#include <list>     ///
#include <stack>    ///
#include <vector>  /// for std::vector
#include <cassert>  /// for assert
#include <limits>   ///
```

**Comment by sauravUppoor** [comment]
@Panquesito7 I have added a short description for the header files.

File: graph/depth_first_search_with_stack.cpp [hunk 14]
```diff
@@ -1,46 +1,207 @@
...
+ * 3. Once a node is popped out of the stack and is coloured GREY, we colour it BLACK.
+ * 4. Push all its neighbours which are not coloured BLACK.
+ * 5. Repeat steps 4 and 5 until the stack is empty.
+ */
+
+#include <iostream> /// for IO operations
+#include <stack> /// header for std::stack
+#include <vector> /// header for std::vector
+#include <cassert> /// header for preprocessor macro assert() 
+#include <limits> /// header for limits of integral types
```
Comment (Panquesito7) [inline graph/depth_first_search_with_stack.cpp:36]:
```suggestion
#include <iostream> /// for IO operations
#include <stack>    /// header for std::stack
#include <vector>   /// header for std::vector
#include <cassert>  /// header for preprocessor macro assert() 
#include <limits>   /// header for limits of integral types
```

File: graph/depth_first_search_with_stack.cpp [hunk 15]
```diff
@@ -1,46 +1,207 @@
...
+            checked[act] = BLACK;  /// Node has been explored
         }
     }
+    return traversed_path;
 }
+}  // namespace depth_first_search
+}  // namespace graph
+
+/**
+ * Self-test implementation
```
Comment (Panquesito7) [inline graph/depth_first_search_with_stack.cpp:120]:
```suggestion
 * Self-test implementations
```

**Review by Panquesito7 — APPROVED** [review]
Amazing work, I like how you've improved the code so well.
Code and documentation looks good and very refined; LGTM. 😄 👍

**Review by ayaankhan98 — APPROVED** [review]
LGTM

**Pull request merged**